### PR TITLE
RetroAchievements - Patch Allowlist Unit Test

### PR DESCRIPTION
**Pull request by LillyJadeKatrin**
This unit test compares ApprovedInis.json with the contents of the GameSettings folder to verify that every patch marked allowed for use with RetroAchievements has a hash in ApprovedInis.json. If not, that hash is reported in the test logs so that the hash may be updated more easily.

### CODE REVIEW

**Comment by AdmiralCurtiss** [comment]
Before you start rewriting the entire test system here, have you considered just embedding the allowed patch list in the executable instead of loading them at runtime?

**Comment by LillyJadeKatrin** [comment]
> Before you start rewriting the entire test system here, have you considered just embedding the allowed patch list in the executable instead of loading them at runtime?

Aside from a general distaste of making design changes purely to facilitate testing, this still has to load all the sys files so I still need to copy them all into the unit tests folder.

**Comment by AdmiralCurtiss** [comment]
Oh right, I guess you still need access to the INIs themselves.

File: Data/Sys/GameSettings/GCCE01.ini [hunk 1]
```diff
@@ -1,5 +1,8 @@
 # GCCE01 - FINAL FANTASY Crystal Chronicles
 
+# This patch is approved for RetroAchievements use. Please see
+# Sys/ApprovedInis.txt for more information.
+
```
Comment (AdmiralCurtiss) [inline Data/Sys/GameSettings/GCCE01.ini:5]: This seems superfluous now.

File: Source/Core/Common/IniFile.cpp [hunk 2]
```diff
@@ -102,6 +102,9 @@ bool IniFile::Section::GetLines(std::vector<std::string>* lines, const bool remo
 
     if (remove_comments)
     {
+      if (*stripped_line.begin() == '@')
+        continue;
+
```
Comment (AdmiralCurtiss) [inline Source/Core/Common/IniFile.cpp:107]: I don't think this does anything useful...?

File: Source/Core/Core/AchievementManager.h [hunk 3]
```diff
@@ -27,7 +27,9 @@
 #include "Common/CommonTypes.h"
 #include "Common/Event.h"
 #include "Common/HttpRequest.h"
+#include "Common/JsonUtil.h"
 #include "Common/WorkQueueThread.h"
+#include "Core/PatchEngine.h"
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.h:32]: This can be removed with the forward declaration.

File: Source/Core/Core/AchievementManager.h [hunk 4]
```diff
@@ -226,6 +245,10 @@ class AchievementManager
 
 #include <string>
 
+#include "Core/ActionReplay.h"
+#include "Core/GeckoCode.h"
+#include "Core/PatchEngine.h"
+
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.h:251]: These can be removed.

File: Source/Core/Core/ActionReplay.cpp [hunk 5]
```diff
@@ -39,6 +39,7 @@
 #include "Common/MsgHandler.h"
 
 #include "Core/ARDecrypt.h"
+#include "Core/AchievementManager.h"
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/ActionReplay.cpp:42]: This can be removed.

File: Source/Core/Core/GeckoCodeConfig.cpp [hunk 6]
```diff
@@ -13,6 +13,7 @@
 #include "Common/IniFile.h"
 #include "Common/Logging/Log.h"
 #include "Common/StringUtil.h"
+#include "Core/AchievementManager.h"
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/GeckoCodeConfig.cpp:16]: This can be removed.

File: Data/Sys/ApprovedInis.json [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+    "CAB9CED2D904F12CCB21F5B1DE9B5433620C3E13" : "loophack"
+  },
+  "G2BE5G" : {
+		"title" : "Black & Bruised",
+    "7FFF6BDD93713BEDFD23739C32B86153FA19AEA0" : "Disable interlaced rendering"
+  },
+  "G2BP7D" : {
+		"title" : "Black & Bruised",
+    "56E85D7285F10348E1E5354E379918D07E79EDA9" : "Disable interlaced rendering"
+  },
```
Comment (AdmiralCurtiss) [inline Data/Sys/ApprovedInis.json:13]: Something is very messy with the formatting in this file. Please run it through a prettifier or something like that.

File: Source/Core/Core/AchievementManager.cpp [hunk 8]
```diff
@@ -64,12 +66,44 @@ void AchievementManager::Init()
...
+    LoadApprovedList();
     if (HasAPIToken())
       Login("");
     INFO_LOG_FMT(ACHIEVEMENTS, "Achievement Manager Initialized");
   }
 }
 
+void AchievementManager::LoadApprovedList()
+{
+  m_queue.EmplaceItem([this]() {
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:78]: This is a race condition on `m_ini_root`, you cannot guarantee that the load is done by the time `FilterApprovedPatches()` is called. Personally I would just move the load of the json into the constructor to work around that...

Comment (LillyJadeKatrin) [reply Source/Core/Core/AchievementManager.cpp:78]: Not sure I understand how putting it in the constructor would fix that, but okay.

Comment (AdmiralCurtiss) [reply Source/Core/Core/AchievementManager.cpp:77]: To be clear: Don't put it in the queue, just load it synchronously.

File: Source/Core/Core/AchievementManager.h (hunk 4 above)
Comment (LillyJadeKatrin) [reply Source/Core/Core/AchievementManager.h:250]: Not sure what happened here because these popped back in a couple times but FYI they've been removed and for some reason GitHub isn't marking this as Outdated.

Comment (AdmiralCurtiss) [reply Source/Core/Core/AchievementManager.h:250]: Nope, they're still there.

File: Source/Core/Common/Crypto/SHA1.h [hunk 9]
```diff
@@ -24,6 +24,12 @@ class Context
 public:
   virtual ~Context() = default;
   virtual void Update(const u8* msg, size_t len) = 0;
+  void Update(const std::vector<u8>& msg) { return Update(msg.data(), msg.size()); }
+  template <size_t length>
+  void Update(const std::array<u8, length>& data)
+  {
+    return Update(data.data(), data.size());
+  }
```
Comment (AdmiralCurtiss) [inline Source/Core/Common/Crypto/SHA1.h:32]: You can remove these, the `std::span` overload just handles them.

Comment (LillyJadeKatrin) [reply Source/Core/Common/Crypto/SHA1.h:32]: I looked it up and wasn't convinced, guess I should have just tried it.

File: Source/Core/Core/AchievementManager.h (hunk 4 above)
Comment (LillyJadeKatrin) [reply Source/Core/Core/AchievementManager.h:250]: Oh my lord, I wasn't paying attention to the line numbers. Visual Studio, why are you like this?

File: Source/Core/Core/AchievementManager.cpp [hunk 10]
```diff
@@ -322,6 +352,48 @@ bool AchievementManager::IsHardcoreModeActive() const
...
+void AchievementManager::FilterApprovedPatches(std::vector<PatchEngine::Patch>& patches,
+                                               const std::string& game_ini_id) const
+{
+  if (!IsHardcoreModeActive())
+    return;
+
+  if (!m_ini_root.contains(game_ini_id))
+    patches.clear();
+  auto patch_itr = patches.begin();
+  while (patch_itr < patches.end())
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:364]: `patch_itr != patches.end()` is the more correct way to check for reaching the end of an iterator.

File: Source/Core/Core/AchievementManager.cpp [hunk 11]
```diff
@@ -322,6 +352,48 @@ bool AchievementManager::IsHardcoreModeActive() const
...
+      context->Update(Common::BitCastToArray<u8>(entry.value));
+      context->Update(Common::BitCastToArray<u8>(entry.comparand));
+      context->Update(Common::BitCastToArray<u8>(entry.conditional));
+    }
+    auto digest = context->Finish();
+
+    bool verified = m_ini_root.get(game_ini_id).contains(Common::SHA1::DigestToString(digest));
+    if (!verified)
+    {
+      patches.erase(patch_itr);
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:383]: This invalidates `patch_itr`. For this kind of loop, you want to do `patch_itr = patches.erase(patch_itr)`. See https://en.cppreference.com/w/cpp/container/vector/erase

**Comment by AdmiralCurtiss** [comment]
Your last commit is weird. You update the allow list but don't update the corresponding hash, which means either the state before or the state after is wrong. You should probably just squash the last and the first commit together anyway, it doesn't seem useful to have a half-finished allow list in here.

**Comment by AdmiralCurtiss** [comment]
Your second commit also has weird in-between edits, for that matter, like the `@Approved` in the FFCC ini.

File: Source/UnitTests/CMakeLists.txt [hunk 12]
```diff
@@ -1,5 +1,14 @@
 enable_testing()
 add_custom_target(unittests)
+
+if(APPLE)
+  set(EXECUTABLE_PATH build/arm64/Source/UnitTests)
+else()
+  set(EXECUTABLE_PATH build/Source/UnitTests)
+endif()
+add_custom_command(TARGET unittests POST_BUILD
+  COMMAND ${CMAKE_COMMAND} -E copy_directory "${CMAKE_SOURCE_DIR}/Data/Sys" "${CMAKE_SOURCE_DIR}/${EXECUTABLE_PATH}/Sys"
```
Comment (AdmiralCurtiss) [inline Source/UnitTests/CMakeLists.txt:10]: I'm... not sure what you're trying to do here. What is `EXECUTABLE_PATH` and why are you so sure of its relationship to the source directory? Do you just want `CMAKE_BINARY_DIR` or what...?

File: .gitignore [hunk 13]
```diff
@@ -41,3 +41,4 @@ CMakeLists.txt.user
 /.idea/
 # Ignore Visual Studio Code's working dir
 /.vscode/
+/Source/UnitTests/Sys
```
Comment (AdmiralCurtiss) [inline .gitignore:44]: This screams at me that you're doing something wrong. A build should never affect the source directory.

File: Source/Core/Core/AchievementManager.cpp [hunk 14]
```diff
@@ -70,6 +72,34 @@ void AchievementManager::Init()
...
+  {
+    WARN_LOG_FMT(ACHIEVEMENTS, "Failed to verify approved game settings list {}",
+                 APPROVED_LIST_FILENAME);
+    WARN_LOG_FMT(ACHIEVEMENTS, "Expected hash {}, found hash {}",
+                 Common::SHA1::DigestToString(APPROVED_LIST_HASH),
+                 Common::SHA1::DigestToString(digest));
+    return;
+  }
+  std::lock_guard lg{m_lock};
+  m_ini_root = temp;
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:100]: `m_ini_root = std::move(temp);`, probably.

File: Source/UnitTests/Core/PatchAllowlistTest.cpp [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+      // Report: approved list has entries when ini has none approved
+      ADD_FAILURE() << "Game in list has no approved patches." << std::endl
+                    << "Game ID: " << game_itr->first << ":" << game_itr->second.game_title;
+    }
+    if (game_itr == allow_list.end() && patches.size() > 0)
+    {
+      // Report: approved patches were found in ini but not in approved list
+      ADD_FAILURE() << "Game with approved patches not found in list." << std::endl
+                    << "Game ID: " << game_id;
+    }*/
```
Comment (AdmiralCurtiss) [inline Source/UnitTests/Core/PatchAllowlistTest.cpp:83]: ?

File: .gitignore (hunk 13 above)
Comment (LillyJadeKatrin) [reply .gitignore:44]: I don't particularly like this either but I haven't figured out what's going wrong with how I'm doing this and haven't gotten any responses when I've asked about it. The weird problem is that when the two Windows autobuilds build and run this unit test, they run the following
```  std::string cur_directory = File::GetCurrentDir();
  std::string sys_directory = cur_directory + DIR_SEP + "Sys";
  const auto& list_filepath = fmt::format("{}{}{}", sys_directory, DIR_SEP, APPROVED_LIST_FILENAME);
  ASSERT_TRUE(JsonFromFile(list_filepath, &json_tree, &error))
      << "Failed to open file at " << list_filepath;```
and the resulting error I get is that list_filepath is looking in _Source_/UnitTests/Sys for the file. I have no idea why the auto builds would attempt to do this and what to do to get them to go somewhere more reasonable, like Binary/x64/Release/UnitTests/Sys which is where the other autobuilds that use CMake are looking.

File: Source/UnitTests/UnitTests.vcxproj [hunk 16]
```diff
@@ -50,7 +56,6 @@
     <ClCompile Include="Common\EventTest.cpp" />
     <ClCompile Include="Common\FileUtilTest.cpp" />
     <ClCompile Include="Common\FixedSizeQueueTest.cpp" />
-    <ClCompile Include="Common\FlagTest.cpp" />
```
Comment (AdmiralCurtiss) [inline Source/UnitTests/UnitTests.vcxproj:53]: Why did you remove this?

File: .gitignore (hunk 13 above)
Comment (AdmiralCurtiss) [reply .gitignore:44]: Okay, so that just means the working directory is set wrong/weird for the unit test executable invokation. Let's see if I can figure out where that is set...

File: Source/UnitTests/UnitTests.vcxproj (hunk 16 above)
Comment (LillyJadeKatrin) [reply Source/UnitTests/UnitTests.vcxproj:53]: Not a clue how this one happened O_o It didn't show up in Visual Studio as a change either.

File: Source/UnitTests/CMakeLists.txt (hunk 12 above)
Comment (AdmiralCurtiss) [reply Source/UnitTests/CMakeLists.txt:10]: In fact, do you just want:

```
add_custom_command(TARGET unittests POST_BUILD
  COMMAND ${CMAKE_COMMAND} -E copy_directory "${CMAKE_SOURCE_DIR}/Data/Sys" "${CMAKE_CURRENT_BINARY_DIR}/Sys"
)
```

?

File: .gitignore (hunk 13 above)
Comment (AdmiralCurtiss) [reply .gitignore:44]: Don't forget to remove this now that we have a better solution.
